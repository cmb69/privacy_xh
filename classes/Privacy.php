<?php

/**
 * Copyright (C) Christoph M. Becker
 *
 * This file is part of Privacy_XH.
 *
 * Privacy_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Privacy_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Privacy_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Privacy;

use Privacy\Infra\Html;
use Privacy\Infra\Newsbox;
use Privacy\Infra\Request;
use Privacy\Infra\View;
use Privacy\Value\Response;

class Privacy
{
    /** @var array<string,string> */
    private $conf;

    /** @var Newsbox */
    private $newsbox;

    /** @var View */
    private $view;

    /** @param array<string,string> $conf */
    public function __construct(array $conf, Newsbox $newsbox, View $view)
    {
        $this->conf = $conf;
        $this->newsbox = $newsbox;
        $this->view = $view;
    }

    public function __invoke(Request $request): Response
    {
        if ($request->adm()) {
            return Response::create();
        }
        switch ($request->privacyAction()) {
            default:
                return $this->show($request);
            case "consent":
                return $this->consent($request);
            case "decline":
                return $this->decline($request);
        }
    }

    private function show(Request $request): Response
    {
        if ($request->isCookieSet() && !$request->showPrivacy()) {
            return Response::create($this->renderLink($request->privacyformUrl()));
        }
        return Response::create($this->renderForm());
    }

    private function consent(Request $request): Response
    {
        return Response::redirect($request->privacyRedirectUrl())
            ->withCookie("privacy_agreed", "yes", $this->getExpirationTime($request->time()));
    }

    private function decline(Request $request): Response
    {
        return Response::redirect($request->privacyRedirectUrl())
            ->withCookie("privacy_agreed", "no", 0);
    }

    private function renderForm(): string
    {
        return $this->view->render("privacy", [
            "message" => $this->conf["newsbox"] !== ""
                ? Html::of($this->newsbox->content($this->conf["newsbox"]))
                : null,
        ]);
    }

    private function renderLink(string $url): string
    {
        return $this->view->render("link", [
            "url" => $url,
        ]);
    }

    private function getExpirationTime(int $now): int
    {
        return (int) $this->conf["duration"] > 0
            ? $now + 24 * 60 * 60 * (int) $this->conf["duration"]
            : 0;
    }
}
