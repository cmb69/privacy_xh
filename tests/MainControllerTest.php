<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Privacy_XH.
 *
 * Privacy_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Privacy_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Privacy_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Privacy;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Privacy\Infra\Request;
use Privacy\Infra\View;

class MainControllerTest extends TestCase
{
    public function testRendersPrivacyForm(): void
    {
        $_POST = [];
        $sut = new MainController(0, $this->message(), $this->view());
        $response = $sut($this->request());
        Approvals::verifyHtml($response->output());
    }

    public function testDoesNothingIfAlreadyAgreed(): void
    {
        $_COOKIE = ["privacy_agreed" => "yes"];
        $sut = new MainController(0, $this->message(), $this->view());
        $response = $sut($this->request());
        $this->assertEquals("", $response->output());
    }

    public function testUserAgrees(): void
    {
        $_POST = ["privacy_agree" => "yes"];
        $sut = new MainController(1, $this->message(), $this->view());
        $request = $this->request();
        $request->method("queryString")->willReturn("some+query+string");
        $response = $sut($request);
        $this->assertEquals(["privacy_agreed", "yes", 86400], $response->cookie());
        $this->assertEquals("http://example.com?some+query+string", $response->location());
    }

    public function testUserDisagrees(): void
    {
        $_POST = ["privacy_agree" => "no"];
        $sut = new MainController(0, $this->message(), $this->view());
        $response = $sut($this->request());
        $this->assertEquals(["privacy_agreed", "no", 0], $response->cookie());
        $this->assertEquals("http://example.com", $response->location());
    }

    private function request(): Request
    {
        return $this->createMock(Request::class);
    }

    private function message(): string
    {
        return XH_includeVar("./languages/en.php", "plugin_tx")["privacy"]["message"];
    }

    private function view(): View
    {
        return new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["privacy"]);
    }
}
