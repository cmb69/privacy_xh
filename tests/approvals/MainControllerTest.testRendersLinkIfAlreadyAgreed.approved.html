<!-- privacy link -->
<p><a href="/?&amp;privacy_show">Cookie settings</a></p>
